research/picai/nnunet_scripts/transfer_train.py

```diff
@@ -0,0 +1,277 @@
+        - The finetuning dataset must be properly formatted and in the nnUNet_raw folder
+        - A plans file for the pretrained model must be present in the nnUNnet_preprocessed
+            under the folder of the dataset it was pretrained on.
+        - The dimensions and voxel spacings of the finetuning dataset must
+            be the same as the pretraining dataset
+        - The pretrained model must already be saved and stored as a pytorch ckpt
+"""
+
+import argparse
+import os
+import shutil
+from multiprocessing import Process
+
+import torch
+from batchgenerators.utilities.file_and_folder_operations import load_json, save_json
+from nnunetv2.experiment_planning.plan_and_preprocess_api import extract_fingerprints, preprocess_dataset
+from nnunetv2.experiment_planning.plans_for_pretraining.move_plans_between_datasets import move_plans_between_datasets
+from nnunetv2.paths import nnUNet_preprocessed, nnUNet_raw
+from nnunetv2.run.run_training import run_training
+from nnunetv2.utilities.dataset_name_id_conversion import convert_id_to_dataset_name
+
+
+def setup_device(device: str) -> torch.device:
+    """Checks the device being used, sets the number of threads and returns a pytorch device
+
+    Args:
+        device: Either 'cpu', 'cuda' or 'mps'
+
+    Returns
+        torch_device: The corresponding pytorch.device object
+    """
+    assert device in [
+        "cpu",
+        "cuda",
+        "mps",
+    ], f"-device must be either cpu, mps or cuda. Other devices are not tested/supported. Got: {device}."
+    if device == "cpu":
+        # let's allow torch to use hella threads
+        import multiprocessing
+
+        torch.set_num_threads(multiprocessing.cpu_count())
+        return torch.device("cpu")
+    elif device == "cuda":
+        # multithreading in torch doesn't help nnU-Net if run on GPU
+        torch.set_num_threads(1)
+        torch.set_num_interop_threads(1)
+        return torch.device("cuda")
+    else:
+        return torch.device("mps")
+
+
+def transfer_metadata(ft_id: int, pt_id: int, pt_plans_identifier: str) -> str:
+    """Transfers and or modifies metadata necessary for transfer learning
+
+    Ensures the dataset json, dataset fingerprint and nnunet plans are all
+    ready for transfer learning. Some necessary changes are made to the plans
+    from pretraining to allow finetuning to occur without issue. Note that
+    this function saves the plans for finetuning as 'pt_plans_identifier-ftXptY'
+    where X is ft_if and Y is pt_id. If a plan file is already found that
+    matches this pattern, the function does not overwrite it.
+
+    Args:
+        ft_id: The nnunet id of the finetuning dataset
+        pt_id: The nnunet id of the pretraining dataset
+        pt_plans_identifier: The identifier/name of the plans from pretraining
+            to use as a basis for the finetuning plans
+    Returns:
+        tl_plans_identifier: The identifier for the plans to use during finetuning
```

Comment:
If we are using tl (transfer learning) and ft (finetuning) interchangeably, maybe we should stick to using one or the other

Reply:
maybe we can expand out pt to pretrain and ft to finetune (if thats what we go with) in the variable names below too